crates/bevy_pbr/src/render/pbr_lighting.wgsl

```diff
@@ -280,3 +280,151 @@ fn directional_light(light_id: u32, roughness: f32, NdotV: f32, normal: vec3<f32
+    // exit positions. IRL, light rays can be scattered when entering/exiting a material (due to
+    // roughness) or inside the material (due to subsurface scattering). Here, we only consider
+    // the first scenario.
+    //
+    // Blur intensity is:
+    // - squarely proportional to `perceptual_roughness`
+    // - inversely proportional to object distance to the view
+    let blur_intensity = (perceptual_roughness * perceptual_roughness) / distance;
+
+    // Number of taps scale with blur intensity
+    let num_taps = i32(max(min(sqrt(blur_intensity) * 7.0, 1.0) * f32(MAX_TRANSMISSIVE_TAPS), 1.0));
+    let num_spirals = (num_taps >> 3u) + 1;
+    let random_angle = interleaved_gradient_noise(frag_coord.xy);
+#ifdef TAA
+    // Alternating pixel mesh: 0 or 1 on even/odd pixels, alternates every frame
+    let pixel_mesh = (i32(frag_coord.x) + i32(frag_coord.y) + i32(globals.frame_count)) % 2;
+#else
+    // Alternating pixel mesh: 0 or 1 on even/odd pixels
+    let pixel_mesh = (i32(frag_coord.x) + i32(frag_coord.y)) % 2;
+#endif
+
+    var result = vec4<f32>(0.0);
+    for (var i: i32 = 0; i < num_taps; i = i + 1) {
+        let current_spiral = (i >> 3u);
+        let angle = (random_angle + f32(current_spiral) / f32(num_spirals)) * 2.0 * PI;
+        let m = vec2(sin(angle), cos(angle));
+        let rotation_matrix = mat2x2(
+            m.y, -m.x,
+            m.x, m.y
+        );
+
+        // Get spiral offset from array
+        var spiral_offset: vec2<f32>;
+        switch i & 7 {
+            // TODO: Figure out a more reasonable way of doing this, as WGSL
+            // seems to only allow constant indexes into the constant array
+            case 0: { spiral_offset = SPIRAL_OFFSETS[0]; }
+            case 1: { spiral_offset = SPIRAL_OFFSETS[1]; }
+            case 2: { spiral_offset = SPIRAL_OFFSETS[2]; }
+            case 3: { spiral_offset = SPIRAL_OFFSETS[3]; }
+            case 4: { spiral_offset = SPIRAL_OFFSETS[4]; }
+            case 5: { spiral_offset = SPIRAL_OFFSETS[5]; }
+            case 6: { spiral_offset = SPIRAL_OFFSETS[6]; }
+            case 7: { spiral_offset = SPIRAL_OFFSETS[7]; }
+            default: {}
+        }
+
+        // Make each consecutive spiral slightly smaller than the previous one
+        spiral_offset *= 1.0 - (0.5 * f32(current_spiral) / f32(num_spirals));
+
+        // Rotate and correct for aspect ratio
+        let rotated_spiral_offset = (rotation_matrix * spiral_offset) * vec2(1.0, aspect);
+
+        // Calucalte final offset position, with blur and spiral offset
+        let modified_offset_position = offset_position + rotated_spiral_offset * blur_intensity * (1.0 - f32(pixel_mesh) * 0.1);
+
+#ifdef PREPASS_TEXTURES
+        // Use depth prepass data to reject values that are in front of the current fragment
+        if (prepass_depth(vec4<f32>(modified_offset_position * view.viewport.zw, 0.0, 0.0), 0u) > frag_coord.z) {
+            continue;
+        }
+#endif
```

Comment:
PREPASS_TEXTURES is true when using webgl with msaa off, but in that case we still can't sample from the depth texture.